### ISSUE TITLE
Migrar framework de segurança

### ISSUE THREAD

**Issue by arthurgregorio** [arthurgregorio/web-budget#231]
Migrar todo o sistema de segurança do PL para outro framework com suporte ao JEE pois o PL esta praticamente parado e sem manutenção/desenvolvimento faz algum tempo.

TL:DR

http://picketlink.org/news/2015/03/10/PicketLink-and-Keycloak-project-merge/

Possível substituto: Apache Shiro.

**Comment by DaniloAndrade** [comment]
O apache shiro não tem um suporte nativo para jsf, e até onde vi, ele não tem muitas atualizações também. 
Acredito que o PicketLink por estar dentro dos modulos do jboss/wildfly a chance de surgir novas versões é bem maior que o shiro.

**Comment by arthurgregorio** [comment]
Interessante!

Estive olhando apenas a parte de como construir por enquanto, não havia me aprofundado sobre as versões e tudo mais.

Estou construindo um projeto de testes para viabilizar alguma solução, ainda sem previsões, pelo menos por enquanto.

**Comment by DaniloAndrade** [comment]
Não gosto muito de misturar java ee e spring, mas talvez uma opção melhor que o shiro seja o  [spring security](http://projects.spring.io/spring-security/)

**Comment by arthurgregorio** [comment]
Na versão 1.x.x do WB ele usava SS, mas como foi migrado para JEE ele praticamente ficou sem possibilidade de uso visto a integração entre ele e o ecossistema JEE.

Uma outra opção é esperar o JEE 8 e a JSR-375

**Comment by DaniloAndrade** [comment]
o JSR-375 é uma boa opção, mas vai demorar um pouco ainda.

**Comment by DaniloAndrade** [comment]
e acredito que o próprio PicketLink venha ser uma implementação da JSR-375

**Comment by DaniloAndrade** [comment]
Sei la, acho que se o picketLink estiver atendo as necessidades do projeto, não deve ser substituído no momento. E esperar o andamento da spec e do próprio wildfly pra ver pra que caminho eles vão tomar.

**Comment by clebiovieira** [comment]
Amigo, Apache Shiro não deve atender pois ele não tinha ate onde sei, suporte a CDI. Você teria que implementar muita coisa que hoje o picketlink te da com facilidade. Tem um projeto se segurança que é baseado em servidor, vc já deve ter visto o Keycloak. Acho que a abordagem de ter um servidor de autenticação seria uma solução mais "Enterprise".  A única coisa que não gostei no picketlink, foi que não conseguir separar as entidades pra fora do projeto Web na época. Abraços

**Comment by arthurgregorio** [comment]
Olá Clebio, basicamente o shiro não possui suporte direto, mas é possível expor as partes necessárias via metodos produtores. No meu repo de exemplos tem um projeto teste com o shiro + cdi.

Quanto ao KC, ele é uma solução grande de mais para meu projeto, seria como matar uma mosca com um canhão.

**Comment by clebiovieira** [comment]
Início de 2016 eu também usei Apache Shiro como teste. A solução também foi por produtores, esse é o caminho natural então. Você já tentou utilizar o Spring Security?
Abçs 

Att,
Clebio Vieira

**Comment by arthurgregorio** [comment]
Sim, antes da versão 2.x.x do sistema ele usava todo o ecossistema do spring, mas fiz a troca para o JEE por questões de de manutenibilidade e padronização, a unica coisa que eu não contava era que o PL ia ser jogado fora.

**Comment by adelmo00** [comment]
Eu optaria por colocar o Spring Security de volta pela facilidade de uso e a certeza de que não será descontinuado e também pelo suporte da comunidade, e também suporte ao JSF.

**Comment by clebiovieira** [comment]
Verdade, suporte garantido e conhecimento bem difundido.

Att,
Clebio Vieira

**Comment by arthurgregorio** [comment]
Spring não tem suporte nativo ao JSF, também existem alguns pormenores nele que deixam o processo um pouco manual. 

Mas problemática do spring como um todo é o fato de que vc se amarra ao resto do framework, o spring por não possui integração direta com o CDI e prover isso implica em uma série de trade-offs que não são interessantes do ponto de de vista das dependências atuais do projeto.

Basicamente, ou vc utiliza todo o spring ou não utiliza, meio termo disso torna a manutenção ruim. 

Quanto a comunidade/suporte, quem mexe com Spring ou PL saberá utilizar o shiro também, o conceito por trás de todos é o mesmo.

Um bom exemplo de como o Shiro é bem mais interessante é que a própria caelum utiliza ele para autenticação/autorização em sua plataforma Alura.
